### PR TITLE
Compute and Network CLI Commands: Features and bug fixes for Sprint49

### PR DESCRIPTION
**Pull request by anuchandy**
Following issues are addressed in this PR.

#1566
#1600
#1726
#1731
#1759
#1761
#1763
#1764
#1769
#1770
#1771
#1773
#1775
#1776
#1777
#1780
#1781
#1783

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@anuchandy__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (v-ancha). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, or work for Microsoft Open Technologies, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: lib/commands/arm/network/dnsRecordSet._js [hunk 1]
```diff
@@ -73,84 +63,73 @@ __.extend(DnsRecordSet.prototype, {
...
-  },
-
-  list: function (resourceGroupName, zoneName, options, _) {
-    var dnsZoneCrud = new DnsZone(this.cli, this.dnsManagementClient);
-
-    var dnsZone = dnsZoneCrud.get(resourceGroupName, zoneName, _);
 
-    if (!dnsZone) {
-      throw new Error($('DNS zone not found'));
+    if (result.code === 204) {
```
Comment (yugangw-msft) [inline lib/commands/arm/network/dnsRecordSet._js]: should not this logic be handled at SDK?

File: lib/commands/arm/network/dnsRecordSet._js [hunk 2]
```diff
@@ -73,84 +63,73 @@ __.extend(DnsRecordSet.prototype, {
...
-        progress2.end();
+    var progress = this.cli.interaction.progress($('Looking up the DNS record sets'));
+    try {
+      if (options.type) {
+        options.type = this._validateType(options.type);
+        dnsRecords = this.dnsManagementClient.recordSets.list(resourceGroupName, zoneName, options.type, options, _);
+      } else {
+        dnsRecords = this.dnsManagementClient.recordSets.listAll(resourceGroupName, zoneName, options, _);
       }
+    } catch (e) {
```
Comment (yugangw-msft) [inline lib/commands/arm/network/dnsRecordSet._js]: you don't need to catch and rethrow. Just remove

File: lib/commands/arm/network/dnsRecordSet._js (hunk 1 above)
Comment (anuchandy) [inline lib/commands/arm/network/dnsRecordSet._js]: It is suppose to handle in SDK, but we found that is not happening. We ill inform this to @jtuliani

File: lib/commands/arm/network/dnsRecordSet._js [hunk 3]
```diff
@@ -411,50 +280,69 @@ __.extend(DnsRecordSet.prototype, {
...
-  	  if(__.isEmpty(recordSet.properties.soaRecord)) {
-  	  	delete recordSet.properties.soaRecord;
-  	  }
-
-  	  if(__.isEmpty(recordSet.properties.ptrRecords)) {
-  	  	delete recordSet.properties.ptrRecords;
-  	  }
-  	}
+    if (recordSet.properties) {
+      if(__.isEmpty(recordSet.properties.aRecords)) {
```
Comment (yugangw-msft) [inline lib/commands/arm/network/dnsRecordSet._js]: consider to looping on a list of properties so to consolidate all the pretty much same if statements

File: lib/commands/arm/network/network._js [hunk 4]
```diff
@@ -861,11 +863,99 @@ exports.init = function (cli) {
...
+    });
+
+  nicAddressPool.command('remove [resource-group] [name]')
+    .description($('Remove a backend address pool from a NIC'))
+    .option('-g, --resource-group <resource-group>', $('the name of the resource group'))
+    .option('-n, --name <name>', $('the name of the network interface'))
+    .option('-i, --lb-address-pool-id  <lb-address-pool-id>', $('the load balancer address pool identifier' +
+    '\n   e.g. /subscriptions/<subscription-id>/resourceGroups/<resource-group-name>/providers/Microsoft.Network/loadbalancers/<lb-name>/backendAddressPools/<address-pool-name>'))
+    .option('-l, --lb-name <lb-name>', $('the load balancer name.' +
+    '\n   This load balancer must exists in the same resource group as the NIC.' +
```
Comment (yugangw-msft) [inline lib/commands/arm/network/network._js]: "exist**~~s~~**"

File: lib/commands/arm/network/vnet._js [hunk 5]
```diff
@@ -147,7 +147,7 @@ __.extend(VNet.prototype, {
     var progress = this.cli.interaction.progress(util.format($('Deleting virtual network "%s"'), name));
     try {
       this.networkResourceProviderClient.virtualNetworks.deleteMethod(resourceGroup, name, _);
-    } catch(e) {
+    } catch (e) {
```
Comment (yugangw-msft) [inline lib/commands/arm/network/vnet._js]: simple "catch, do nothing and rethrow"  is unnecessary

File: lib/commands/asm/vm/vmclient.js [hunk 6]
```diff
@@ -4948,4 +5133,74 @@ function lookupExtension(extensionRefs, extensionName, publisherName, options) {
...
 
+function parseNICParams(nicParams) {
+  var networkInterfaces = [];
+  var nicConfig = {
+    error: null,
+    networkInterfaces: networkInterfaces
+  };
+
+  var nicConfigFmtMsg = '--nic-config must be comma seperated NIC definitions, format of each NIC definition must be <nicName>:<subnetName>:[vnetStaticIP]:[nsgName]:[ipForwarding]';
+  if (nicParams && (nicParams = nicParams.trim())) {
```
Comment (yugangw-msft) [inline lib/commands/asm/vm/vmclient.js]: could you just use (utils.stringIsNullOrEmpty(nicParams )) so to be consistent with your other code

File: lib/util/logging.js [hunk 7]
```diff
@@ -124,6 +124,43 @@ log.table = function (level, data, transform) {
...
 };
 
+log.nameValue = function (name, value, indent, displayNullValue) {
+  var delimiter = ': ';
+  if (!displayNullValue) displayNullValue = false;
+  if (!indent) indent = 0;
+  var key = spaces(indent) + name;
+  key += spaces(32 - key.length);
+
+  if (value !== undefined && value !== null) {
```
Comment (yugangw-msft) [inline lib/util/logging.js]: if (value)

File: test/commands/arm/network/arm.network.lb-address-pool-tests.js [hunk 8]
```diff
@@ -41,15 +42,15 @@ describe('arm', function () {
 	describe('network', function () {
     var suite,
 	    retry = 5;
```
Comment (yugangw-msft) [inline test/commands/arm/network/arm.network.lb-address-pool-tests.js]: please fix the indent and not aligned whitespaces

**Comment by yugangw-msft** [comment]
please pull from the latest dev and get CI pass

File: lib/commands/asm/vm/vmclient.js (hunk 6 above)
Comment (anuchandy) [inline lib/commands/asm/vm/vmclient.js]: Ok, will use

if (utils.stringIsNullOrEmpty(nicParams )) {
  nicParams =  nicParams.trim();

File: test/commands/arm/network/arm.network.lb-address-pool-tests.js (hunk 8 above)
Comment (anuchandy) [inline test/commands/arm/network/arm.network.lb-address-pool-tests.js]: cc @babumuralidharan

File: lib/commands/arm/network/network._js (hunk 4 above)
Comment (anuchandy) [inline lib/commands/arm/network/network._js]: @yugangw-msft  sorry, you mean put the word -exists- inside double quotes?

File: lib/util/logging.js (hunk 7 above)
Comment (anuchandy) [inline lib/util/logging.js]: we allow empty string in-case of nameValue func.

File: lib/commands/arm/network/dnsRecordSet._js (hunk 3 above)
Comment (anuchandy) [inline lib/commands/arm/network/dnsRecordSet._js]: Ok, will do like:

var recordTypes = ['aRecords', 'aaaaRecords', .., ..., ..,];
for (var i = 0; i < recordTypes.length; i++) {
  if(__.isEmpty(recordSet.properties[recordTypes[i]])) {
    delete recordSet.properties[recordTypes[i]];
  }
}

File: lib/commands/arm/network/network._js (hunk 4 above)
Comment (ruslangabitov) [inline lib/commands/arm/network/network._js]: Fixed this issue here https://github.com/MSOpenTech/azure-xplat-cli/commit/e40708ab2619b1cc2732c0548040265f5687be68
Don't know why this comment is not shown as outdated.

File: test/commands/arm/network/arm.network.dnszone-tests.js [hunk 9]
```diff
@@ -20,114 +19,95 @@ var util = require('util');
...
+    tag = 'priority=medium;size=high';
 var requiredEnvironment = [{
     name: 'AZURE_VM_TEST_LOCATION',
     defaultValue: 'southeastasia'
-    }];
+}];
+
+
+describe('arm', function() {
+    describe('network', function() {
```
Comment (yugangw-msft) [inline test/commands/arm/network/arm.network.dnszone-tests.js]: the indent should be 2 spaces for xplat repo. Please update them

File: lib/commands/arm/network/network._js (hunk 4 above)
Comment (yugangw-msft) [inline lib/commands/arm/network/network._js]: i mean "exist" without 's'

Comment (yugangw-msft) [inline lib/commands/arm/network/network._js]: and looks like it has been addressed

**Pull request merged**